app/components/Portfolio/Modal.tsx

```diff
@@ -0,0 +1,230 @@
+import { AssetData } from "@/app/store/slices/portfolioSlice";
+import Image from "next/image";
+import * as Icons from "@/app/icons";
+import { getTodayDate } from "@/app/utils/utils";
+import { v4 as uuidv4 } from "uuid";
+
+type ModalProps = {
+  isOpen: boolean;
+  onClose: () => void;
+  onUpdateAssets: (newAsset: AssetData) => void;
+  assetToEdit?: AssetData;
+  onDeleteAsset: (assetId: string) => void;
+};
+
+const Modal: React.FC<ModalProps> = ({
+  isOpen,
+  onClose,
+  onUpdateAssets,
+  assetToEdit,
+  onDeleteAsset,
+}) => {
+  const coins = useAppSelector((state) => state.coinsData);
+  const dispatch = useAppDispatch();
+  const [selectedCoin, setSelectedCoin] = useState<CoinData | null>(null);
+  const [purchasedAmount, setPurchasedAmount] = useState(0);
+  const [selectedDate, setSelectedDate] = useState("");
+
+  // change state for asset to edit
+  useEffect(() => {
+    if (assetToEdit) {
+      setSelectedCoin(
+        coins.find((coin) => coin.name === assetToEdit.name) || null
+      );
+      setPurchasedAmount(assetToEdit.total_value);
+      setSelectedDate(assetToEdit.purchase_date.substring(0, 10));
+    } else {
+      clearModalState();
+    }
+  }, [assetToEdit, coins]);
+
+  const clearModalState = () => {
+    setSelectedCoin(null);
+    setPurchasedAmount(0);
+    setSelectedDate("");
+  };
+
+  const handleClose = () => {
+    clearModalState();
+    onClose();
+  };
+
+  const handleSelectChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
+    const selectedCoinId = e.target.value;
+    const matchedCoin = coins.find((coin) => coin.id === selectedCoinId);
+    if (matchedCoin) {
+      setSelectedCoin(matchedCoin);
+    }
+  };
+
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const value = e.target.value;
+    if (parseFloat(value) > 0) {
+      setPurchasedAmount(value ? parseFloat(value) : 0);
+    } else {
+      setPurchasedAmount(0);
+    }
+  };
+
+  const handleDateChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setSelectedDate(e.target.value);
+  };
+
+  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    if (selectedCoin && purchasedAmount > 0 && selectedDate) {
+      try {
+        // dispatch thunk and wait for profit percentage result
+        const profitPercentage = await dispatch(
+          fetchProfitPercentage({
+            assetId: selectedCoin.id,
+            purchaseDate: new Date(selectedDate + "T00:00:00").toISOString(),
+            currentPrice: selectedCoin.current_price,
+          })
+        ).unwrap();
+
+        // Finalized asset to be created or edited
+        const newAsset: AssetData = {
+          id: assetToEdit ? assetToEdit.id : uuidv4(), // edited asset's id or generate new one
+          symbol: selectedCoin.symbol,
+          name: selectedCoin.name,
+          image: selectedCoin.image,
+          total_value: purchasedAmount,
```

Comment:
you could also get rid og the snake case if you are creating a new object

Reply:
Converted all to camelCase.